File: Front_Coelsa/src/domain/validadorCuit.ts
// Espejo de Coelsa.Domain/Validaciones/ValidadorCuit.cs (módulo 11).

const PESOS = [5, 4, 3, 2, 7, 6, 5, 4, 3, 2] as const;

/** Dígito verificador módulo 11: resto 0 => 0, resto 1 => 9, resto n => 11 - n. */
export function calcularDigitoVerificador(primerosDiezDigitos: string): number {
  let suma = 0;
  for (let i = 0; i < 10; i++) {
    suma += Number(primerosDiezDigitos[i]) * PESOS[i];
  }
  const resto = suma % 11;
  return resto === 0 ? 0 : resto === 1 ? 9 : 11 - resto;
}

export function esCuitValido(cuit: string | null | undefined): boolean {
  if (!cuit || cuit.length !== 11 || !/^\d{11}$/.test(cuit)) {
    return false;
  }
  return calcularDigitoVerificador(cuit.slice(0, 10)) === Number(cuit[10]);
}

/** Completa una base de 10 dígitos con su dígito verificador (útil para seeds y tests). */
export function completarCuit(baseDiezDigitos: string): string {
  return baseDiezDigitos + String(calcularDigitoVerificador(baseDiezDigitos));
}

File: IrmaRios/Identidad/src/Identidad.Domain/Validaciones/ValidadorCuit.cs
using System.Text.RegularExpressions;

namespace IrmaRios.Identidad.Domain.Validaciones;

/// <summary>
/// Validación de CUIT/CUIL compartida con el simulador COELSA (mismo algoritmo módulo 11).
/// </summary>
public static partial class ValidadorCuit
{
    public const int Longitud = 11;

    [GeneratedRegex(@"^\d{11}$")]
    private static partial Regex PatronCuit();

    private static readonly int[] Pesos = [5, 4, 3, 2, 7, 6, 5, 4, 3, 2];

    public static bool EsValido(string? cuit)
    {
        if (cuit is null || cuit.Length != Longitud || !PatronCuit().IsMatch(cuit))
        {
            return false;
        }

        return CalcularDigitoVerificador(cuit[..10]) == cuit[10] - '0';
    }

    /// <summary>Dígito verificador módulo 11: resto 0 =&gt; 0, resto 1 =&gt; 9, resto n =&gt; 11 - n.</summary>
    public static int CalcularDigitoVerificador(string primerosDiezDigitos)
    {
        var suma = 0;
        for (var i = 0; i < 10; i++)
        {
            suma += (primerosDiezDigitos[i] - '0') * Pesos[i];
        }

        var resto = suma % 11;
        return resto == 0 ? 0 : resto == 1 ? 9 : 11 - resto;
    }

    /// <summary>Completa una base de 10 dígitos con su dígito verificador (útil para seeds y tests).</summary>
    public static string Completar(string baseDiezDigitos)
    {
        return baseDiezDigitos + CalcularDigitoVerificador(baseDiezDigitos);
    }
}
